tikv/raftstore/snap_manager.go

```diff
@@ -0,0 +1,355 @@
+package raftstore
+
+import (
+	"github.com/ngaut/log"
+	"github.com/pingcap/errors"
+	rspb "github.com/pingcap/kvproto/pkg/raft_serverpb"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type SnapEntry int
+
+const (
+	SnapEntryGenerating SnapEntry = 1
+	SnapEntrySending    SnapEntry = 2
+	SnapEntryReceiving  SnapEntry = 3
+	SnapEntryApplying   SnapEntry = 4
+)
+
+func (e SnapEntry) String() string {
+	switch e {
+	case SnapEntryGenerating:
+		return "generating"
+	case SnapEntrySending:
+		return "sending"
+	case SnapEntryReceiving:
+		return "receiving"
+	case SnapEntryApplying:
+		return "applying"
+	}
+	return "unknown"
+}
+
+type SnapStats struct {
+	ReceivingCount int
+	SendingCount   int
+}
+
+func notifyStats(ch chan<- *Msg) {
+	if ch != nil {
+		ch <- &Msg{Type: MsgTypeStoreSnapshotStats}
+	}
+}
+
+type SnapManager struct {
+	base         string
+	snapSize     *int64
+	registryLock sync.RWMutex
+	registry     map[SnapKey][]SnapEntry
+	ch           chan<- *Msg
+	limiter      *IOLimiter
+	MaxTotalSize uint64
+}
+
+func NewSnapManager(path string, ch chan<- *Msg) *SnapManager {
+	return new(SnapManagerBuilder).Build(path, ch)
+}
+
+func (sm *SnapManager) init() error {
+	fi, err := os.Stat(sm.base)
```

Comment:
Should acquire write lock to prevent multiple threads call init at the same time.

Reply:
I suppose SnapManger.init() would only be called once.

Reply:
OK